OpenSim/Common/Test/testStorage.cpp

```diff
@@ -27,77 +27,113 @@
-                    "testStorage_version2_short_header.sto");
-            Storage stVersion2("testStorage_version2_short_header.sto");
-            SimTK_TEST_EQ(table.getMatrix(),
-                    stVersion2.exportToTable().getMatrix());
-        }
+
+void testStorageLoadingFromFile(const std::string& fileName, const int ncols);
+
+void testStorageLegacy() {
+    // Create a storage from a std file "std_storage.sto"
+    //ofstream checkRunDirFile("rundir.txt");
+    //checkRunDirFile << "Run from here:\n\n";
+    //checkRunDirFile.close();
+    string stdLabels[] = { "time", "v1", "v2" };
+    std::unique_ptr<Storage> st(new Storage("test.sto"));
+    // time[\t]v1[\t]v2
+    // 1.[\t]   10.0[Space]20
+    // 2.[\t\t] 20.0[\t]40
+    ASSERT(st->getSize() == 2);
+    const Array<std::string> &lbls = st->getColumnLabels();
+    ASSERT(lbls.getSize() == 3);
+    int i = 0;
+    for (i = 0; i<lbls.getSize(); i++) {
+        ASSERT(lbls[i] == stdLabels[i]);
+    }
+
+    double val;
+    for (i = 0; i<st->getSize(); i++) {
+        StateVector& row = (*st->getStateVector(i));
+        ASSERT(row.getTime() == i + 1);
+        ASSERT(row.getData()[0] == row.getTime()*10.0);
+        row.getDataValue(0, val);
+        ASSERT(val == row.getTime()*10.0);
+        ASSERT(row.getData()[0] == row.getTime()*10.0);
+        ASSERT(row.getData()[1] == row.getTime()*20.0);
+    }
+    int ncol = st->getSmallestNumberOfStates();
+    ASSERT(ncol == 2);
+    Array<double> col(SimTK::CNT<SimTK::Real>::getNaN(), 4);
+    st->getDataColumn(1, col);
+    ASSERT(col[0] == 20.);
+    ASSERT(col[1] == 40.0);
+
+    ASSERT(st->getStateIndex("v2") == 1);
+
+    Storage st2("testDiff.sto");
+    // Test Comparison
+    double diff = st->compareColumn(st2, stdLabels[1], 0.);
+    ASSERT(fabs(diff) < 1E-7);
+    diff = st->compareColumn(st2, stdLabels[2], 0.);
+    ASSERT(fabs(diff) < 1E-7);
+
+    // Loading version 2 storage file with Storage class.
+    auto table = st->exportToTable();
+    STOFileAdapter::write(table, "testStorage_version2.sto");
+    {
+        // Now read using Storage() constructor.
+        Storage stVersion2("testStorage_version2.sto");
+        // Compare with st.
+        SimTK_TEST_EQ(table.getMatrix(),
+            stVersion2.exportToTable().getMatrix());
     }
-    catch (const Exception& e) {
-        e.print(cerr);
-        return 1;
+
+    // The version 2 storage file does not require nRows and nColumns
+    // metadata (Issue #2120).
+    {
+        table.removeTableMetaDataKey("nRows");
+        table.removeTableMetaDataKey("nColumns");
+        STOFileAdapter::write(table,
+            "testStorage_version2_short_header.sto");
+        Storage stVersion2("testStorage_version2_short_header.sto");
+        SimTK_TEST_EQ(table.getMatrix(),
+            stVersion2.exportToTable().getMatrix());
+    }
+}
+
+int main() {
+    SimTK_START_TEST("testStorage");
+
+        // Verify the loading scalar Outputs (2) from ..sto into a Storage
```

Comment:
It's not clear to me what `(2)` means.

Reply:
That was a note to expect two Outputs. I will clarify.